File: BreadthFirstSearch/200_Number_of_Islands/200_Number_of_Islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) 
    {
        if(grid.empty())
            return 0; 
        
        int res = 0;
        size_t rows = grid.size();
        size_t cols = grid[0].size();
        set<pair<size_t, size_t>> visited;

        for(size_t r = 0; r < rows; r++) {
            for(size_t c = 0; c < cols; c++) {
                auto pos = visited.find(make_pair(r,c));
                if(grid[r][c] == '1' && pos == visited.end()) {
                    bfs(r, c, grid, visited, rows, cols);
                    res++;
                }
            }
        }
        return res;    
    }
private:
    // do not pass visited by value, which means that changes to visited inside bfs will not be reflected outside the function. 
    // You should pass it by reference instead.
    void bfs(size_t r, size_t c, vector<vector<char>>& grid, set<pair<size_t, size_t>>& visited, size_t rows, size_t cols) 
    {
        vector<pair<int, int>> directions {{1, 0},{-1, 0},{0, 1},{0, -1}};
        queue<pair<size_t, size_t>> q;
        visited.insert(make_pair(r,c));
        q.push(make_pair(r,c));
        
        while(!q.empty()) {
            pair<size_t, size_t> curr = q.front(); q.pop();
            size_t row = curr.first;
            size_t col = curr.second;
            for(auto p : directions) {
                int dr = p.first;
                int dc = p.second;
                size_t curr_r = row + dr;
                size_t curr_c = col + dc;
                auto pos = visited.find(make_pair(curr_r,curr_c));
                if(curr_r < rows && curr_c < cols && grid[curr_r][curr_c] == '1' && pos == visited.end()) {
                    q.push(make_pair(curr_r, curr_c));
                    visited.insert(make_pair(curr_r, curr_c));
                }
            }
        }
    }
};

File: BreadthFirstSearch/200_Number_of_Islands/200_Number_of_Islands.java
import java.util.*;

class Solution {
    public int numIslands(char[][] grid) {

         if (grid.length < 1 || grid[0].length < 1) {
            return 0; //empty
        }
        
        int res = 0;
        int rows = grid.length;
        int cols = grid[0].length;
        Set<Map.Entry<Integer, Integer>> visited = new HashSet<>();
        
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                Map.Entry<Integer, Integer> pair = new AbstractMap.SimpleEntry<>(r, c);
                if (grid[r][c] == '1' && !visited.contains(pair)) {
                    bfs(r, c, grid, visited, rows, cols);
                    res++;
                }
            }
        }
        return res;    
    }
    private void bfs(int r, int c, char[][] grid, Set<Map.Entry<Integer, Integer>> visited, int rows, int cols) {
        List<Map.Entry<Integer, Integer>> directions = Arrays.asList(
            new AbstractMap.SimpleEntry<>(1, 0),
            new AbstractMap.SimpleEntry<>(-1, 0),
            new AbstractMap.SimpleEntry<>(0, 1),
            new AbstractMap.SimpleEntry<>(0, -1)
        );
        
        Queue<Map.Entry<Integer, Integer>> q = new LinkedList<>();
        visited.add(new AbstractMap.SimpleEntry<>(r, c));
        q.add(new AbstractMap.SimpleEntry<>(r, c));
        
        while (!q.isEmpty()) {
            Map.Entry<Integer, Integer> curr = q.poll();
            int row = curr.getKey();
            int col = curr.getValue();
            for (Map.Entry<Integer, Integer> p : directions) {
                int dr = p.getKey();
                int dc = p.getValue();
                int curr_r = row + dr;
                int curr_c = col + dc;
                Map.Entry<Integer, Integer> pair = new AbstractMap.SimpleEntry<>(curr_r, curr_c);
                if (curr_r >= 0 && curr_r < rows && curr_c >= 0 && curr_c < cols && grid[curr_r][curr_c] == '1' && !visited.contains(pair)) {
                    q.add(pair);
                    visited.add(pair);
                }
            }
        }
    }
}